File: dotnet/DocumentIntelligence.Api/Services/QualityDetector.cs
namespace DocumentIntelligence.Api.Services;

public enum DocumentQualityTier
{
    DigitalPdf,
    ScannedPdf,
    Photo
}

public static class QualityDetector
{
    private static readonly HashSet<string> ImageExtensions = ["jpg", "jpeg", "png", "tiff", "bmp", "webp"];
    private const int DigitalPdfTextThreshold = 500;

    public static DocumentQualityTier Detect(string filename, string? extractedText)
    {
        var ext = Path.GetExtension(filename).TrimStart('.').ToLowerInvariant();

        if (ImageExtensions.Contains(ext))
            return DocumentQualityTier.Photo;

        if (ext == "pdf")
        {
            return extractedText?.Trim().Length >= DigitalPdfTextThreshold
                ? DocumentQualityTier.DigitalPdf
                : DocumentQualityTier.ScannedPdf;
        }

        return DocumentQualityTier.Photo;
    }
}

File: app/services/quality_detector.py
from enum import Enum


class DocumentQualityTier(str, Enum):
    DIGITAL_PDF = "digital_pdf"
    SCANNED_PDF = "scanned_pdf"
    PHOTO = "photo"


_IMAGE_EXTENSIONS = {"jpg", "jpeg", "png", "tiff", "bmp", "webp"}
_DIGITAL_PDF_TEXT_THRESHOLD = 500


def detect_quality(file_bytes: bytes, filename: str, extracted_text: str | None) -> DocumentQualityTier:
    ext = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""

    if ext in _IMAGE_EXTENSIONS:
        return DocumentQualityTier.PHOTO

    if ext == "pdf":
        text = extracted_text or ""
        if len(text.strip()) >= _DIGITAL_PDF_TEXT_THRESHOLD:
            return DocumentQualityTier.DIGITAL_PDF
        return DocumentQualityTier.SCANNED_PDF

    return DocumentQualityTier.PHOTO
